src/relay/pass/merge_composite.cc

```diff
@@ -0,0 +1,192 @@
+    Expr empty_expr;
+    if (!pattern->op->IsInstance<OpNode>() || !root->op->IsInstance<OpNode>())
+      return empty_expr;
+    if (pattern->op.as<OpNode>()->name != root->op.as<OpNode>()->name)
+      return empty_expr;
+    if (pattern->args.size() != root->args.size())
+      return empty_expr;
+
+    unsigned int i = 0;
+    Array<Expr> new_args;
+    for (const auto& arg : pattern->args) {
+      if (arg->IsInstance<CallNode>()) {
+        new_args.push_back(ExtractPattern(Downcast<Call>(arg),
+                                          Downcast<Call>(root->args[i]),
+                                          var_map));
+      }
+      if (arg->IsInstance<VarNode>()) {
+        new_args.push_back(ExtractPattern(Downcast<Var>(arg),
+                                          root->args[i],
+                                          var_map));
+      }
+      if (arg->IsInstance<ConstantNode>()) {
+        new_args.push_back(ExtractPattern(Downcast<Constant>(arg),
+                                          root->args[i],
+                                          var_map));
+      }
+      i++;
+    }
+
+    auto new_call = CallNode::make(root->op, new_args, root->attrs);
+    return new_call;
+  }
+
+  Expr VisitExpr_(const CallNode* cn) {
+    Call call = GetRef<Call>(cn);
+    if (call->op->IsInstance<FunctionNode>()) {
+      Function func = Downcast<Function>(call->op);
+      CHECK(func.defined());
+      const auto name_node = FunctionGetAttr(func, attr::kComposite).as<tir::StringImmNode>();
+      if (name_node->value != "") {
+        tvm::Array<tvm::relay::Expr> new_args;
+        for (const auto& arg : call->args) {
+          auto new_e = this->Mutate(arg);
+          new_args.push_back(new_e);
+        }
+        return CallNode::make(call->op, new_args, call->attrs);
+      }
+    }
+
+    Expr expr = ExprMutator::VisitExpr_(cn);
+    call = Downcast<Call>(expr);
+    if (!call->op->IsInstance<OpNode>())
+      return call;
+
+    Op op = Downcast<Op>(call->op);
+    CHECK(op.defined());
+    for (const auto& x : pattern_map_) {
+      Call pattern = Downcast<Call>(x.second);
+      if (Downcast<Op>(pattern->op)->name != op->name)
+        continue;
+
+      if (MatchPattern(pattern, call)) {
+        Map<std::string, Array<Expr>> args_map;
+        auto extract = ExtractPattern(pattern, call, &args_map);
+        auto free_vars = FreeVars(extract);
+        Function new_func = FunctionNode::make(free_vars, extract,
+                call->checked_type_, {}, Attrs());
+        new_func = FunctionSetAttr(new_func, attr::kComposite,
+                                   tir::StringImmNode::make(x.first));
+        new_func = FunctionSetAttr(new_func, attr::kPrimitive,
+            tvm::Integer(1));
+        Array<Expr> args;
+        for (const auto& free_var : free_vars) {
+          args.push_back(args_map[free_var->name_hint()][1]);
+        }
+        auto new_call = CallNode::make(new_func, args);
+        return new_call;
```

Comment:
Does this mean if there multiple patterns which can be matched to this calll, it will return the first match as determined by tvm::Map key order?

Reply:
Same question here. If this PR does not attempt to deal with multiple pattern matching to one call, it would be better to provide a priority number for users to specify which pattern should be matched prior to others.

Reply:
Yep, this is an oversight. I can fix it pretty easily to work in priority order though and shall update the PR accordingly. I'll add tests for this case.